cpp/ql/lib/semmle/code/cpp/security/InvalidPointerDereference/InvalidPointerToDereference.qll

```diff
@@ -2,6 +2,69 @@
  * This file provides the second phase of the `cpp/invalid-pointer-deref` query that identifies flow
  * from the out-of-bounds pointer identified by the `AllocationToInvalidPointer.qll` library to
  * a dereference of the out-of-bounds pointer.
+ *
+ * Consider the following snippet:
+ * ```cpp
+ * 1. char* base = (char*)malloc(size);
+ * 2. char* end = base + size;
+ * 3. for(int *p = base; p <= end; p++) {
+ * 4.   use(*p);
+ * 5. }
+ * ```
+ * this file identifies the flow from `base + size` to `end`. We call `base + size` the "dereference source" and `end`
+ * the "dereference sink" (even though `end` is not actually dereferenced - it will be used to find the correct
+ * dereference eventually).
+ *
+ * Merely _constructing_ a pointer that's out-of-bounds is fine if the pointer is never dereferenced (in reality, the
+ * standard only guarentees that it's safe to move the pointer one element past the last element. But we ignore that
+ * here). So this step is about identifying which of those out-of-bounds pointers identified from step 1 that are
+ * actually being dereferenced. We do this using a regular dataflow configuration (see `InvalidPointerToDerefConfig`).
+ *
+ * This dataflow traversal defines the set of sources as any dataflow node that is non-strictly lower-bounded by the
+ * pointer-arithmetic instruction identified by `AllocationToInvalidPointer.qll`. That is, the set of sources is any
+ * dataflow node `source` such that `source.asInstruction() >= pai + delta1` for some `delta1 >= 0`.
```

Comment:
I would stick to just writing a formula, as we're now doing in `AllocationToInvalidPointer`. Also, we might need to update `delta1` to some more sensible name? Same for `delta2` below.

Reply:
Good point. Fixed in 55cfadb1f4bff2919c129f0127372dde1d9c705c. I'll fix `delta2` in a separate commit.